src/shared/components/Animations/InfiniteScroll.tsx

```diff
@@ -0,0 +1,125 @@
+import * as React from 'react';
+import { Empty } from 'antd';
+import './infinite-scroll.less';
+import './list-item.less';
+import { useTransition, animated, useSpring, config } from 'react-spring';
+import { PaginatedList } from '@bbp/nexus-sdk';
+import { FetchableState } from '../../store/reducers/utils';
+import useInfiniteScroll from '../hooks/useInfiniteScroll';
+
+const DEFAULT_TRAIL_MS = 50;
+
+const DEFAULT_ANIMATIONS = {
+  from: { transform: 'scale3d(0.8, 0.8, 0.8)', height: 0, opacity: 0 },
+  leave: { transform: 'scale3d(0.8, 0.8, 0.8)', height: 0, opacity: 0 },
+  enter: { transform: 'scale3d(1, 1, 1)', height: 'auto', opacity: 1 },
+  trail: DEFAULT_TRAIL_MS,
+};
+
+export const InfiniteScrollLoadMoreButton: React.FunctionComponent<{
+  hasMore: boolean;
+  totalItemsListLength: number;
+  onClick: VoidFunction;
+  isFetching: boolean;
+}> = ({ hasMore, totalItemsListLength, onClick, isFetching }) => {
+  const animatedOpacity = useSpring({
+    opacity: hasMore ? 1 : 0,
+    // TODO why: This doesn't work as described https://www.react-spring.io/docs/hooks/api
+    config: config.molasses,
+    delay: DEFAULT_TRAIL_MS * totalItemsListLength,
+  });
+  const animatedLoadTransition = useTransition(isFetching, null, {
+    from: { position: 'absolute', opacity: 0 },
+    enter: { opacity: 1 },
+    leave: { opacity: 0 },
+  });
+  return (
+    <animated.div style={animatedOpacity}>
+      <a onClick={onClick}>
+        <li className="list-item -action -load">
+          <div className="loading">
+            {animatedLoadTransition.map(({ item, props }) =>
+              item ? (
+                <animated.div style={props} className="center-helper">
+                  <div className="center">Loading</div>
+                </animated.div>
+              ) : (
+                <animated.div style={props} className="center-helper">
+                  <div className="center">Load More</div>
+                </animated.div>
+              )
+            )}
+          </div>
+        </li>
+      </a>
+    </animated.div>
+  );
+};
+
+export interface InfiniteScrollProps {
+  itemComponent: (item: any, index: number) => React.ReactElement;
+  itemClassName?: string;
+  makeKey: (item: any, index: number) => string;
+  next: VoidFunction;
+  style?: React.CSSProperties;
+  fetchablePaginatedList: FetchableState<PaginatedList<any>>;
+}
+
+const InfiniteScroll: React.FunctionComponent<InfiniteScrollProps> = props => {
+  const {
+    makeKey,
```

Comment:
what does this do?

Reply:
this is a map function so you can take the item and return a value to make a string for the key, useful I think

Reply:
Another reason why we need this is because the elements are animated, and those animated react elements must host the key, not the item component thats rendered inside it, so we need a way to map the keys inside the InfiniteScroll component itself